### ISSUE TITLE
Graph Transform Tool fails to optimize away useless Switch node

### ISSUE THREAD

**Issue by jchia** [tensorflow/tensorflow#23992]
**System information**
- TensorFlow version (you are using): v1.12.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
The Graph Transform Tool apparently fails to optimize away Switch nodes that have constant pred input, as demonstrated in the test case. The test program shows a GraphDef that includes an unnecessary Switch node that could have been optimized away.

When a Switch node has pred input that is constant, it should be possible to optimize the node away but currently this optimization does not happen. Consequently, when using the Graph Transform Tool to optimize a GraphDef for a network that is built to be used for training, eval and inference, that has a bool placeholder to differentiate the use case, nodes needed only for training remain even when the placeholder node has been converted to a constant False node and 'fold_constants' is specified to the Graph Transform Tool. Together with 'strip_unused_nodes', the desired optimization would result in other nodes also being removed that are used/available only during training.

This optimization could be provided as a new transform or as part of 'fold_constants'.

Test:
```
#!/usr/bin/env python3

import tensorflow as tf
from tensorflow.tools.graph_transforms import TransformGraph

with tf.Graph().as_default() as g:
    pred = tf.constant(False, dtype=tf.bool, shape=(), name='pred')

    def one(): return tf.constant(1, dtype=tf.int32, shape=(), name='one')

    def zero(): return tf.constant(0, dtype=tf.int32, shape=(), name='zero')

    r = tf.cond(pred, one, zero)

gd = g.as_graph_def()
transforms = ['strip_unused_nodes', 'fold_constants', 'strip_unused_nodes']
gd = TransformGraph(gd, [], [r.name], transforms)
print(gd)
```

**Will this change the current api? How?**
No

**Who will benefit with this feature?**
Especially users who build one graph for both training and inference and differentiate the use case using a boolean placeholder that controls the behavior of the batch norm or dropout in the graph. This feature would help them get cleaner and smaller graphs for inference.

**Any Other info.**

**Comment by jchia** [comment]
This is related to #5919 but from one of the comments, I suppose that was for `tensorflow.python.tools.optimize_for_inference_lib`, not the Graph Transform Tool.

**Comment by WIll-Xu35** [comment]
I have the same problem.

After optimized the model for deployment using Graph Transform Tool. Armnn raises error for this optimized model. After checking the network structure, I found all the Switch op are untouched. And the constant value input (which is of type 'DT_BOOL') to all the Switches are causing the trouble.

After that, I tried to add 'op=Switch' to 'remove_nodes()', but the result pb file is exactly the same as before.

Apart from waiting for a new release, is there any workaround for now?

**Comment by msymp** [comment]
Hello @jchia , your suggestion for optimization as part of the Graph Transform Tool for Switch nodes is useful.
We currently don't have anybody working on this. It would be great if you could help us by working on this and submitting a PR. 
We will close this issue now. Thanks!
